x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts

```diff
@@ -465,26 +465,45 @@ export type Threats = t.TypeOf<typeof threats>;
 export const threatsOrUndefined = t.union([threats, t.undefined]);
 export type ThreatsOrUndefined = t.TypeOf<typeof threatsOrUndefined>;
 
-export const threshold = t.intersection([
-  t.exact(
-    t.type({
-      field: t.union([t.string, t.array(t.string)]),
-      value: PositiveIntegerGreaterThanZero,
-    })
-  ),
-  t.exact(
-    t.partial({
-      cardinality_field: t.union([t.string, t.array(t.string), t.undefined, t.null]),
-      cardinality_value: t.union([PositiveInteger, t.undefined, t.null]), // TODO: cardinality_value should be set if cardinality_field is set
-    })
-  ),
-]);
-// TODO: codec to transform threshold field string to string[] ?
+export const thresholdField = t.exact(
+  t.type({
+    field: t.union([t.string, t.array(t.string)]), // Covers pre- and post-7.12
+    value: PositiveIntegerGreaterThanZero,
+  })
+);
+export type ThresholdField = t.TypeOf<typeof thresholdField>;
+
+export const thresholdFieldNormalized = t.exact(
+  t.type({
+    field: t.array(t.string),
+    value: PositiveIntegerGreaterThanZero,
+  })
+);
+export type ThresholdFieldNormalized = t.TypeOf<typeof thresholdFieldNormalized>;
+
+export const thresholdCardinalityField = t.exact(
+  t.partial({
+    cardinality_field: t.array(t.string),
+    cardinality_value: PositiveInteger,
```

Comment:
The create rules route with this schema won't check that both `cardinality_field` and `cardinality_value` are required, but the rule would fail validation when executing due to the checks [here](https://github.com/elastic/kibana/pull/92667/files#diff-e32ea750ee430426fc82d7c1336eac9d931328b8542f466f47b59b28e3a64076R51).

We can avoid the extra validation functions if we change the schema to be
```
t.exact(
  t.partial({
    cardinality: t.exact(t.type({
      field: t.array(t.string),
      value: PositiveInteger,
    })),
  }),
)
```
The extra validation function appears correct but it's harder to follow imo and the logic will need to be duplicated everywhere we define the schemas. With this schema `cardinality` is optional but if provided it must have both `field` and `value`, and both io-ts and joi can represent this without custom validators.